targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/device/TOOLCHAIN_GCC_ARM/MPS2.ld

```diff
@@ -17,21 +14,20 @@
  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
- */
-/*
+ *
  * This file is derivative of CMSIS V5.00 gcc_arm.ld
+ *
+ * Linker script for mbed CM3DS on MPS2
  */
-/* Linker script for mbed CM3DS on MPS2 */
 
-/* Linker script to configure memory regions. */
-/* The length of the VECTORS region is a bit larger than
- * is necessary based on the number of exception handlers.
- */
+#include "../memory_zones.h"
+#include "../cmsis_nvic.h"
```

Comment:
like all includes in this PR, not using relative paths, `#include "memory_zones.h"`

Reply:
I had to add the relative path because during the preprocessing step of the linker script, all folders are not included in the search path as it is done during compilation.

When I use `#include "memory_zones.h"`, the following command generates an error during mbed OS compilation process (compiling a random test as example):
```
arm-none-eabi-cpp -E -P /fake_path/BUILD/tests/ARM_CM3DS_MPS2/GCC_ARM/targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/device/TOOLCHAIN_GCC_ARM/MPS2.ld -Wl,--gc-sections -Wl,--wrap,main -Wl,--wrap,_malloc_r -Wl,--wrap,_free_r -Wl,--wrap,_realloc_r -Wl,--wrap,_memalign_r -Wl,--wrap,_calloc_r -Wl,--wrap,exit -Wl,--wrap,atexit -Wl,-n -mcpu=cortex-m3 -mthumb -o /fake_path/BUILD/tests/ARM_CM3DS_MPS2/GCC_ARM/./TESTS/mbed_hal/ticker/.link_script.ld
```
Adding `-I /fake_path/BUILD/tests/ARM_CM3DS_MPS2/GCC_ARM/targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/device/`, which is the folder containing `memory_zones.h`, at the end of the command resolves the issue. It is the same story for `ARM` compiler.

**Looking at it a bit closer:**
For `GCC_ARM`, the `link` function in `tools/toolchains/gcc.py` preprocess the linker script before calling the link command (using `preproc_output` path). I guess the preprocessing command could be enhanced with the path of every folder?
For `ARM` compiler, the preprocessor is executed on the linker script thanks to the shebang at the beginning of the file so I could add the include path there. However the `correct_scatter_shebang` function in `tools/toolchains/arm.py` will change it if it is not beginning with the correct format. This function could also add the path of every folder in the command.

I am happy to raise an issue about it :smiley:

Reply:
Also, I had to rebase for the code to compile after #6229 was merged.

Reply:
Thanks for the info. I was not aware linker does not have the paths.

@theotherjimmy What do you think? If a linker file needs an include, should it work with absolute paths?

Reply:
It should **never** use absolute paths. They are not only not portable, but also not ever portable between users!

-------

@0xc0170 You probably meant something other than absolute paths.